File: 03/day03.pl
:- ['../lib/io.pl'].

trees(XIncr, YIncr, Ans) :-
    maxX(MaxX), maxY(MaxY),
    numlist(1, MaxY, Ys),
    include([X]>>(X mod YIncr #= 0), Ys, YList),
    maplist([Y,X-Y]>>(X#=(XIncr*(Y//YIncr)) mod MaxX), YList, Coords),
    foldl([X-Y,Old,New]>>(tree(X,Y)->New#=Old+1;New=Old), Coords, 0, Ans).

part1(Ans) :-
    trees(3, 1, Ans).

part2(Ans) :-
    trees(1, 1, R1D1),
    trees(3, 1, R3D1),
    trees(5, 1, R5D1),
    trees(7, 1, R7D1),
    trees(1, 2, R1D2),
    Ans #= R1D1 * R3D1 * R5D1 * R7D1 * R1D2.

assert_trees(Y, Line) :-
    enumerate(Line, Enumerated),
    forall(member(X-C, Enumerated), (
        char_code(Char, C),
        (Char = '.' ; (Char = '#', assertz(tree(X,Y))))
    )).

parse(Y) -->
    string_without("\n", Line), blanks, eos,
    {assert_trees(Y, Line),
    length(Line, MaxX), assertz(maxX(MaxX)),
    assertz(maxY(Y))}.

parse(Y) -->
    string_without("\n", Line), blanks,
    {assert_trees(Y, Line), NY#=Y+1},
    parse(NY).

run :-
    input_stream(3, parse(0)),
    part1(Ans1),
    write_part1(Ans1),
    part2(Ans2),
    write_part2(Ans2).

File: 03/day03.go
package main

import (
    "github.com/deosjr/adventofcode2020/lib"
)

type coord struct {
    x int
    y int
}

func checkTrees(m map[coord]struct{}, maxx, maxy, incrx, incry int) int {
    trees := 0
    pc := coord{0,0}
    for {
        y := pc.y + incry
        if y >= maxy {
            break
        }
        x := (pc.x + incrx) % maxx
        pc = coord{x:x, y:y}
        if _, ok := m[pc]; ok {
            trees++
        }
    }
    return trees
}

func main() {
    maxx, maxy := 0, 0
    m := map[coord]struct{}{}
    readfunc := func(line string) {
        maxx = len(line)
        for x, c := range line {
            if c != '#' {
                continue
            }
            m[coord{x:x, y:maxy}] = struct{}{}
        }
        maxy++
    }
    lib.ReadFileByLine(3, readfunc)

    r3d1 := checkTrees(m, maxx, maxy, 3, 1)
    lib.WritePart1("%d", r3d1)

    r1d1 := checkTrees(m, maxx, maxy, 1, 1)
    r5d1 := checkTrees(m, maxx, maxy, 5, 1)
    r7d1 := checkTrees(m, maxx, maxy, 7, 1)
    r1d2 := checkTrees(m, maxx, maxy, 1, 2)
    p2 := r1d1 * r3d1 * r5d1 * r7d1 * r1d2
    lib.WritePart2("%d", p2)
}
